packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -695,10 +695,31 @@ export default function parse(source: string, state: State): TemplateParseResult
             }
 
             const { name, location } = attr;
+
             if (!isValidHTMLAttribute(element.tag, name)) {
                 warnAt(ParserDiagnostics.INVALID_HTML_ATTRIBUTE, [name, tag], location);
             }
 
+            // disallow attr name with leading and trailing underscores
+            if (name.startsWith('_') || name.endsWith('_')) {
+                const node = element.__original as parse5.AST.Default.Element;
+                warnAt(ParserDiagnostics.ATTRIBUTE_CANNOT_START_OR_END_WITH_UNDERSCORE, [
+                    name,
+                    treeAdapter.getTagName(node),
+                ]);
+                return;
+            }
+
+            // disallow attr name with an underscore combined with non alphabetic characters
+            if (name.match(/_[^a-zA-Z]|[^a-zA-Z]_/)) {
```

Comment:
since attributes are parsed to lowercase by parse5, you only need to match `a-z` here